### DISCUSSION TITLE
Bundle uninstalls Msi's on upgrade

### DISCUSSION THREAD

**Discussion by tm-nti** [wixtoolset/issues#7606] [Questions]
I am very unsure about this question as I have no idea what is happening

The problem is upgrading from wix v3 created bundle to wix v4 created bundle
When trying to upgrade a bundle "2024.0.000" (made with wix v3) with bundle "2024.0.001" (made with wix4) all the unchanged msi packages are uninstalled (during the BA Previous Versions stage thing)
Now some of these msi packages are third party so we cannot change their version

The problem only happens on upgrade. Running "Repair" afterwards does install the msi correctly or "uninstall/reinstall" also works correctly
Our Product Owner requires upgrade to work, otherwise I have to go back to wix 3 (I wanna stay, dont make me go back)

Has anyone had any similar experience and/or fixes to this problem?

~~Now to make it more confusing~~
~~I have tested this problem and can only make it happen on our "2024" major version. "2023" works fine, but they use the same wix bundle project, the only difference is the version and upgrade code~~

~~Other people on the team have had this issue happen on the "2023" version~~

Logs:
First install, wix v3: [out000.log](https://gist.github.com/tm-nti/af56249524682d7e9504e6fe608bf8e9),
First upgrade, also wix v3: [out002.log](https://gist.github.com/tm-nti/97f74679ea7e454b6df0154736f2eae0)
Second Upgrade, wix v4: [out003.log](https://gist.github.com/tm-nti/d4e166c64458cb6bc19989fcd0ec6935)

**Comment by kerrywicks** [comment]
I think you'd have to set Permanent="yes" attribute on your MsiPackage in your bundle. This would prevent it being uninstalled in an upgrade *and* uninstall scenario though.

**Reply by tm-nti** [reply]
I still need to be able to uninstall it

**Comment by barnson** [comment]
Logs will likely be instructive.

**Reply by tm-nti** [reply]
Good argument, I have updated the description

**Comment by BMurri** [comment]
I'm going to guess that this is somehow related to the change in the default package dependency key (where v3 burn appears to be uninstalling packages even though v4 burn registered dependencies on them).

**Reply by tm-nti** [reply]
Do you know if there is any documentation for these so called package dependency keys? I don't really know what they are or how they are used. And google is not helping me much

**Comment by tm-nti** [comment]
quick update about the extra confusion, turns out I was a dum dum
I thought i was installing a wix 3 made 2023 package and then a wix 4 made package
it was actually a wix 4 made package upgrading a wix 4 made package, so of course it worked fine

**Comment by tm-nti** [comment]
Update
I was looking at the version in "add or remove programs" and it appears that the version string changed between v3 and v4?
![image](https://github.com/wixtoolset/issues/assets/90686450/0d351c9f-a299-4a59-891b-c501ef49a55d)
The two on the left are made with wix 3 and the one on the right is made with wix 4 (top is msi, bottom is bundle)

the format of the version string is unchanged between the wix project upgrade, it comes from Directory.Build.props -> DefineConstants
the format is "24.0.004" for msi and "2024.0.004" for bundle

Does this have any effect on the dependency counter thing?

**Comment by tm-nti** [comment]
It might have looked me in the eyes all along
in wix 3:
`Registering package dependency provider: {3410F465-8107-4844-8CD3-BF4290C5CFDE}, version: 24.1.000, package: NTIFUNCTIONBROWSER.2024`
in wix 4:
`Registering package dependency provider: {3410F465-8107-4844-8CD3-BF4290C5CFDE}_v24.1.000, version: 24.1.000, package: NTIFUNCTIONBROWSER.2024`

now my question is. How do I make it drop the "_v24.1.000"? I have looked at the [Provides element](https://wixtoolset.org/docs/schema/wxs/provides/) but I dont know how I can bind it to the ProductCode of the msi package

**Comment by tm-nti** [comment]
I managed to drop the "_v24.1.000" by temporarily writing the key to a Provides element, but it still did not change anything.

Does anyone know what changed between v3 and v4? I am seriously at wits end

**Comment by BMurri** [comment]
What we really need to see are the logs that would (might?) have been generated by the wix3 bundle that was being uninstalled. Especially comparing it to the previous wix3 bundle that was uninstalled by that wix3 bundle.

Not to say that the bug is in wix3, but to see why it's responding differently.

I honestly don't know if those logs were generated or not, but if they were, they were probably placed either in the same directory as the logs you shared or in/under %TMP%.

**Comment by tm-nti** [comment]
Hello again
I apologise for the sudden silence, but in the meantime I have created a minimal repo with the error
https://github.com/tm-nti/WixShenanigans
See the Readme for instructions

**Reply by mrzade** [reply]
Hello,
could you let me know if you solved the issue?

**Reply by BMurri** [reply]
Requested logs, please (wix3 bundle uninstall log set)

**Comment by mrzade** [comment]
@tm-nti , @BMurri
Hello,
in my case, I have a Wix3 bundle with an MSI package. Wix5's successor upgrades the Wix3 bundle. Both share the same upgrade code.
The MSI package in question remains the same in both bundles.

What does happen is:
- Wix5 successfully detects the Wix3 bundle as a related bundle,
- As expected, Wix5 does not plan any actions to the MSI in question because it has not changed between the bundles.
- Wix5 executes uninstall for the Wix3 bundle after a successful upgrade. The Burn protocol command line arguments seem to be correct.
- Unexpectedly the Wix3's bundle uninstalls the MSI package in question!

Now what is very strange:
1. When upgrading Wix3 with Wix3 everything works as expected and the MSI in question remains present after the upgrade.
2. When upgrading Wix5 with Wix5 everything also works as expected and the MSI remains present after the upgrade.

It looks to be the issue only during the Wix3 -> Wix5 (and I assume Wix4) upgrade.

**Reply by BMurri** [reply]
We need the wix3 bundle uninstall logs, ideally for both the wix3->wix3 and wix3->wix5 scenarios.

I'm pretty sure that wix4 & wix5 will be the same in these scenarios.

**Reply by mrzade** [reply]
Here we go
[test_v3-v3.zip](https://github.com/wixtoolset/issues/files/15106050/test_v3-v3.zip)
[test_v3-v5.zip](https://github.com/wixtoolset/issues/files/15106051/test_v3-v5.zip)
[test_v5-v5.zip](https://github.com/wixtoolset/issues/files/15106052/test_v5-v5.zip)

The MSI in question Pkg.FifSupport.
I don't know how to capture the bundle uninstall log for the Wix3->Wix3 upgrade scenario since Burn doesn't propagate the `-log` switch.

**Reply by mrzade** [reply]
I guess something had changed in the Burn protocol that prevents Wix3 bundles from handling package persistent state correctly. Probably the layout of some structure or the name of a registry key?

**Reply by mrzade** [reply]
Any insights/news?

**Reply by BMurri** [reply]
When burn (of any version) doesn't receive a `-log` switch, it creates a set of logs in the `%TMP%` directory that are named based on the burn package name and timestamp of invocation, among other things. That's what we need to see.

**Reply by mrzade** [reply]
Wix3-Wix3 upgrade #2
[test2_v3-v3.zip](https://github.com/wixtoolset/issues/files/15132529/test2_v3-v3.zip)

I've noticed that during the Wix3-Wix3 upgrade, the previous version of the bundle plans related bundle upgrade (`i207` record of the uninstall log @17:28:43).
This is not the case during Wix3-Wix5 upgrade. Here is the Wix3 bundle plan 6 packages for uninstall (`i200` record @09:34:17).

**Reply by mrzade** [reply]
Any news?

**Reply by El-ke** [reply]
Hello,
I have the same problem with updating from Wix3 to Wix5. 
When I run the wix5 msi to update my dlls and exe, he uninstalls all my dlls but did not install the new one
Is there a solution?

**Reply by metsma** [reply]
You have to add Provides Parameters to packages that are not updated
https://github.com/open-eid/windows-installer/commit/b074d1a6e6c8c6139554d558aa1814246dd20f9b

**Reply by mrzade** [reply]
@metsma 
Hi,
what does the value in the `Key`  attribute mean?

**Reply by metsma** [reply]
It is the UUID of MSI package

**Reply by El-ke** [reply]
Sorry, its not the bootstrapper that makes problem. 
Its the MSI-File. When I run the MSI-File(with wix5) to update the older version (created with wix3), he uninstalls all the files (dlls and .exe) and folders in the installation-folder. If i rerun the MSI, he installs everything correctly.
We only converted the wxs-files from wix3 to wix5.
Have you any idea what we did wrong ?

**Reply by metsma** [reply]
It is the bootstrapper that makes the problem.
If you install with V3 bootstrapper then package is registered as (taken from previous comments here):
`Registering package dependency provider: {3410F465-8107-4844-8CD3-BF4290C5CFDE}, version: 24.1.000, package: NTIFUNCTIONBROWSER.2024`
If you upgrade to V4/5 bootstrapper teh same package is registered with name:
`Registering package dependency provider: {3410F465-8107-4844-8CD3-BF4290C5CFDE}_v24.1.000, version: 24.1.000, package: NTIFUNCTIONBROWSER.2024`
Now on end of upgrade V4 bootstrapper calls V3 bootstrapper and lets him to cleanup all unused msi-s.
Now V3 bootstrapper sees that there is no other providers under "{3410F465-8107-4844-8CD3-BF4290C5CFDE}" and uninstalls the MSI.

The workaround is use in V4/5 bootstrapper Provides attribute
`<Provides Key="{3410F465-8107-4844-8CD3-BF4290C5CFDE}"/>`
It tells to new bootstrapper that register with name "{3410F465-8107-4844-8CD3-BF4290C5CFDE}" and not "{3410F465-8107-4844-8CD3-BF4290C5CFDE}_v24.1.000" another provider.
When on end v3 bootstrapper sees that there is another provider, it will not touch the MSI.

**Reply by El-ke** [reply]
Sorry, 
but I am confused and do not know how to set this <Provides Key…> Element in my MSI-Project. 
I've tried a few things with no success. ☹
Where have the <Providers>-Element be inserted? 
And which Guid should I take? The Guid of the Upgrade-Code?
I can’t find an entry with : “Registering package dependency provider:” in my MSI-Logfile
Can you please give me a hint ? 
Thanks

I have several wxs-files where the dlls and exe-files are defined:
Assembly.wxs
```
<Fragment>
  <!--Not in Gac! If desired add "Assembly =".net"" to all .Netassemblys-->
  <ComponentGroup Id="AgentAssemblys" Directory="INSTALLFOLDER">      
    <Component Id="RemoveAgentInstallableKey" KeyPath="yes" Guid="{6F6B7F67-04AA-4C21-B521-08CAB989FD60}">
      <RemoveRegistryKey Id="RemoveAgentKeyOnUninstall" Action="removeOnUninstall" Root="HKLM" Key="SOFTWARE\XXX\XXX Config\Agent_Installable" />
    </Component>
    <Component Id="Comp_Agent.exe.config" Guid="C1F50C81-BC9D-4618-974A-45F1D92DEE09">
      <RemoveFile Id="Agent.exe.config" Name="Agent.exe.config" On="both" />
      <File Id="Agent.exe.config" Name="Agent.exe.config" KeyPath="yes" Source="$(var.ProjectOutputDir)\Agent.exe.config" />
    </Component>
    <Component Id="Comp_XXX.YYY.Business.DeviceSupport.dll" >
      <File Id="XXX.YYY.Business.DeviceSupport.dll" Name=" XXX.YYY.Business.DeviceSupport.dll" KeyPath="yes" Source="$(var.ProjectOutputDir)\XXX.YYY.Business.DeviceSupport.dll
    </Component>
<!--Note: must be changed in SchedulerAssemblys.wxs as well-->
<ComponentGroup Id="SchedulerAgentSharedAssemblys" Directory="INSTALLFOLDER">
  <Component Id="Comp_XXX.YYY.Business.AgentJobs.dll">
    <File Id="XXX.YYY.Business.AgentJobs.dll" Name="XXX.YYY.Business.AgentJobs.dll" KeyPath="yes" Source="$(var.ProjectOutputDir)\XXX.YYY.Business.AgentJobs.dll"  />
  </Component> 
…

```
DevExpressAssemplies.wxs
```
<Fragment>
  <!--ComponentGroup DevExpressAssemblys-->
  <!--Installed into GAC (Assembly=".net")-->
  <ComponentGroup Id="DevExpressAssemblys" Directory="INSTALLFOLDER">
    <Component Id="Comp_DevExpress.Charts.v22.2.Core.dll">
        **<Provides Key="{2B5BBE5F-DA59-4EF1-B39C-F931B8D4470E}"/> ???**
      <File Id="DevExpress.Charts.v22.2.Core.dll" Name="DevExpress.Charts.v22.2.Core.dll" KeyPath="yes" Assembly=".net" Source="..\..\..\..\3rdPartyAssemblies\DevExpress.Charts.v22.2.Core.dll" />
    </Component>
    <Component Id="Comp_DevExpress.CodeParser.v22.2.dll">
      <File Id="DevExpress.CodeParser.v22.2.dll" Name="DevExpress.CodeParser.v22.2.dll" KeyPath="yes" Assembly=".net" Source="..\..\..\..\3rdPartyAssemblies\DevExpress.CodeParser.v22.2.dll" />
    </Component>…

```

Product.wxs

```
<?include ..\Common\BasePropertys.wxi ?>
 <Package Codepage="1252" Language="1033" Manufacturer="!(loc.Manufacturer)" Name="!(loc.ProductName)" UpgradeCode="{A96CF5EA-2DC8-42EA-B7D1-623A52C5BC1C}" Version="!(bind.FileVersion.Agent.exe)" InstallerVersion="200">
   <SummaryInformation Description="!(loc.ProductName)" Manufacturer="!(loc.Manufacturer)" />

   <!--Condition: NET Framework 4.8 installed -->
   <PropertyRef Id="WIX_IS_NETFRAMEWORK_48_OR_LATER_INSTALLED" />
   <Launch Condition="Installed OR WIX_IS_NETFRAMEWORK_48_OR_LATER_INSTALLED" Message="!(loc.NetError)" />
  …
<InstallExecuteSequence>
  <!--Reference https://docs.microsoft.com/en-us/windows/desktop/msi/suggested-installexecutesequence -->
  <Custom Action="AdjustZeusConfigPath" After="SaveCurrentConfig" Condition="ISFIRSTINSTALL OR ISREPAIR" />
  <Custom Action="SaveCurrentConfig" Before="CostFinalize" Condition="(ISFIRSTINSTALL = &quot;True&quot; OR WIX_UPGRADE_DETECTED)" />

  <Custom Action="StopCurrentServiceWrapper" Before="StopCurrentService" Condition="(ISFIRSTINSTALL = &quot;True&quot; OR WIX_UPGRADE_DETECTED OR ISREPAIR)" />
  <Custom Action="StopCurrentService" After="InstallInitialize" Condition="(ISFIRSTINSTALL = &quot;True&quot; OR WIX_UPGRADE_DETECTED OR ISREPAIR)" />

  <Custom Action="RestoreCurrentConfigWrapper" Before="RestoreCurrentConfig" Condition="NOT ISREMOVE AND ((CAERROR = &quot;0&quot;) AND (ISFIRSTINSTALL = &quot;True&quot; OR WIX_UPGRADE_DETECTED OR ISREPAIR))" />
  <!--On first install or upgrade, not on uninstall-->
  <Custom Action="RestoreCurrentConfig" Before="InstallFinalize" Condition="NOT ISREMOVE AND ((CAERROR = &quot;0&quot;) AND (ISFIRSTINSTALL = &quot;True&quot; OR WIX_UPGRADE_DETECTED OR ISREPAIR))" />
  <!--Prevent change of service user-->
  <DeleteServices Condition="NOT UPGRADINGPRODUCTCODE" />
  <InstallServices Condition="NOT WIX_UPGRADE_DETECTED" />
  <!--Only on upgrade: restart service-->
  <Custom Action="RestartCurrentServiceWrapper" Before="RestartCurrentService" Condition="NOT ISREMOVE AND (WIX_UPGRADE_DETECTED OR ISREPAIR)" />
  <Custom Action="RestartCurrentService" Before="InstallFinalize" Condition="NOT ISREMOVE AND (WIX_UPGRADE_DETECTED OR ISREPAIR)" />
</InstallExecuteSequence>  
…

<Feature Id="ProductFeature" Title="ZeusAgentSetup" Level="1">
  <!--Install DevExpressFiles to Gac-->
  <ComponentGroupRef Id="DevExpressAssemblys" />
  <!--Install other assemblys (not in Gac)-->
  <ComponentGroupRef Id="CommonAssemblys" />
  <ComponentGroupRef Id="CommunicatorAgentSharedAssemblys" />
  <ComponentGroupRef Id="SchedulerAgentSharedAssemblys" />
  <!--Install localization assemblys (not in Gac)-->
  <ComponentGroupRef Id="Localization_en" />
  <ComponentGroupRef Id="Localization_en_us" />
  <ComponentGroupRef Id="Localization_fr" />
  <!--Install agent specific assemblys (not in Gac)-->
  <ComponentGroupRef Id="AgentAssemblys" />
  <!--Install service(s)-->
  <ComponentGroupRef Id="ProductServices" />
</Feature>

```

**Reply by wmanning** [reply]
@El-ke 

Yeah the exact solution above doesn't apply to you since you are using MSI directly and not in a bundle (aka burn).  It might still help.

If you look at the documentation: https://wixtoolset.org/docs/schema/wxs/provides/.
- It tells you the parents the <Provide> element can go under.  For instance "[Component](https://wixtoolset.org/docs/schema/wxs/component/)
Something like:
```
<Component>
    <Provides Key="some-unique-string-id-goes-here" />
</Component>
```

Now the question is what component?  The above documentation link answers that: "It should be authored into a component that is always installed and removed with the product or features that contain it."  Unless you have optional features in your MSI you can add it to any component.

Use it in the MSI that other MSI's need to depend on.

Then in the other MSI installers add a [Requires element](https://wixtoolset.org/docs/schema/wxs/requires/) that references the Key of the above Provides element.  I never used it so I'm only guessing.

I'm not sure if it helps you as you would have to go back to your v3 MSI and add a provides element and release that to customers.  And then update your new MSI to wix v5 and release that.

I always use burn and never ran into your issue with my v3 to v4 upgrade (never upgraded to v5 yet but I think v4 is where the big changes are).  Maybe because I have a simpler scenario than you.

**Reply by wmanning** [reply]
I want to stress I'm just explaining how, I think, it works and how you could use it.  No idea if it'll solve your issue.  You might have an unrelated issue.

**Comment by petm2022** [comment]
The problem sound very much like this issue: https://github.com/wixtoolset/issues/issues/7778
